core/os/flock/flock.go

```diff
@@ -0,0 +1,187 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flock
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+const (
+	fileSuffix = ".gapid-flock"
+	// panic messages
+	acquireAlreadyOwnedLockMsg   = "Try to acquire the lock which is already held by current mutex."
+	unlockAlreadyReleasedLockMsg = "Try to unlock an already released mutex."
+	cannotMakeDirMsg             = "Cannot make temporary directory for flock files."
+)
+
+var dir = filepath.Join(os.TempDir(), "gapid-flocks")
+
+// TryLock creates a Mutex by using a file named with the given string, and
+// tries to acquire the inter-process lock, then returns the created Mutex. This
+// is a non-blocking call. The returned Mutex may not hold the lock, and should
+// be checked with Locked() to tell if the locked is held by the Mutex.
+func TryLock(n string) *Mutex {
+	m := New(n)
+	m.TryLock()
+	return m
+}
+
+// Lock creates a Mutex by using a file named with the given strings, and
+// acquires the inter-process lock, then returns the created Mutex. This is a
+// blocking call and the returned Mutex is guaranteed to hold the lock.
+func Lock(n string) *Mutex {
+	m := New(n)
+	m.Lock()
+	return m
+}
+
+// Mutex is a file based inter-process mutex. Mutex should only be created by
+// New, Lock or TryLock.
+type Mutex struct {
+	fm *mutex
+	tm sync.Mutex
+}
+
+// New creates a new file based inter-process mutex which uses the given string
+// as the name of the underlying file. The Mutex created by New does not hold
+// any lock.
+func New(n string) *Mutex {
+	return &Mutex{
+		fm: &mutex{p: filepath.Join(dir, n+fileSuffix)},
+	}
+}
+
+// Locked returns true if the Mutex holds the lock and valid to call Unlock,
+// false if the Mutex does not hold the lock and valid to call TryLock or Lock.
+// The state of locked/unlocked can only be changed by calling LockFor,
+// TryLockFor and Unlock functions, removing/modifing the underlying files
+// does not change this state.
+func (m *Mutex) Locked() bool {
+	return m.fm.locked
+}
+
+// TryLock acquires the inter-process lock in a non-blocking way. Returns true
+// if the lock is now held by the Mutex, or false if not. Panic if the Mutex
+// already holds the lock.
+func (m *Mutex) TryLock() bool {
+	if m.fm.locked {
+		panic(acquireAlreadyOwnedLockMsg)
+	}
+	m.tm.Lock()
+	defer m.tm.Unlock()
+
+	err := m.fm.tryLock()
+	if err == nil {
+		return true
+	}
+	return false
+}
+
+// Lock acquires the inter-process lock in a blocking way, it only returns
+// when it has acquired the lock. Panic if the Mutex already holds the lock
+// when calling this function.
+func (m *Mutex) Lock() {
+	for {
```

Comment:
You might want to pass down a context and check for cancellation here - this allows the user to interrupt the blocking call.
Can be done in a follow up CL though.

Reply:
Hmm, ok, right. Will add it in another CL.